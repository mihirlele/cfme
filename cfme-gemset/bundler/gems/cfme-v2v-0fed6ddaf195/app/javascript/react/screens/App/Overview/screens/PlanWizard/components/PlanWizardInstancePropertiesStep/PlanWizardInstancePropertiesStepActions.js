import URI from 'urijs';
import API from '../../../../../../../../common/API';

import {
  QUERY_V2V_OSP_TENANT_ATTRIBUTES,
  SET_V2V_INSTANCE_PROPERTIES_ROWS,
  QUERY_V2V_OSP_BEST_FIT_FLAVOR,
  SET_V2V_BEST_FIT_FLAVORS_AND_DEFAULT_SECURITY_GROUPS
} from './PlanWizardInstancePropertiesStepConstants';

export const _getTenantWithAttributesActionCreator = (url, tenantIds) => dispatch =>
  dispatch({
    type: QUERY_V2V_OSP_TENANT_ATTRIBUTES,
    payload: API.post(url, {
      action: 'query',
      resources: tenantIds.map(id => ({ id }))
    })
  });

export const queryTenantsWithAttributesAction = (url, tenantIds, attributes) => {
  const uri = new URI(url);
  uri.addSearch({ expand: 'resources', attributes: attributes.join(',') });
  return _getTenantWithAttributesActionCreator(uri.toString(), tenantIds);
};

export const setFlavorsAndSecurityGroups = (response, mappings, editingPlan, dispatch) => {
  const vmsSlugPrefix = 'vms/';
  const flavorsSlugPrefix = 'flavors/';
  const cloudTenantsSlugPrefix = 'cloud_tenants/';

  const vmBestFitFlavors = [];
  response.data.results.forEach(result => {
    const vmId = result.source_href_slug.slice(vmsSlugPrefix.length);
    const flavorId = result.best_fit && result.best_fit.slice(flavorsSlugPrefix.length);
    const tenantId = mappings
      .find(mapping => mapping.source_href_slug === `${vmsSlugPrefix}${vmId}`)
      .destination_href_slug.slice(cloudTenantsSlugPrefix.length);

    vmBestFitFlavors.push({ vm_id: vmId, tenant_id: tenantId, flavor_id: flavorId });
  });

  dispatch({
    type: SET_V2V_BEST_FIT_FLAVORS_AND_DEFAULT_SECURITY_GROUPS,
    payload: { vmBestFitFlavors, editingPlan }
  });
};

export const _bestFitFlavorActionCreator = (url, flavor_mappings, editingPlan) => dispatch => {
  const postBody = {
    action: 'vm_flavor_fit',
    mappings: flavor_mappings
  };
  return dispatch({
    type: QUERY_V2V_OSP_BEST_FIT_FLAVOR,
    payload: new Promise((resolve, reject) => {
      API.post(url, postBody)
        .then(response => {
          resolve(response);
          setFlavorsAndSecurityGroups(response, flavor_mappings, editingPlan, dispatch);
        })
        .catch(e => {
          reject(e);
        });
    })
  });
};

export const bestFitFlavorAction = (url, mappings, editingPlan) => {
  const uri = new URI(url);
  return _bestFitFlavorActionCreator(uri.toString(), mappings, editingPlan);
};

export const instancePropertiesRowsAction = rows => dispatch =>
  dispatch({
    type: SET_V2V_INSTANCE_PROPERTIES_ROWS,
    payload: rows
  });
