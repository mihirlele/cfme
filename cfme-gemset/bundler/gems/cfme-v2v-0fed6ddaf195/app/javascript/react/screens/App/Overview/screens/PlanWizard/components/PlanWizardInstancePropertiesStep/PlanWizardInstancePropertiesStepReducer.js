import Immutable from 'seamless-immutable';
import {
  QUERY_V2V_OSP_TENANT_ATTRIBUTES,
  SET_V2V_INSTANCE_PROPERTIES_ROWS,
  QUERY_V2V_OSP_BEST_FIT_FLAVOR,
  SET_V2V_BEST_FIT_FLAVORS_AND_DEFAULT_SECURITY_GROUPS,
  PLAN_WIZARD_EXITED,
  PLAN_WIZARD_NEXT
} from './PlanWizardInstancePropertiesStepConstants';

import { stepIDs } from '../../PlanWizardConstants';

export const initialState = Immutable({
  tenantsWithAttributes: [],
  isFetchingTenantsWithAttributes: false,
  isRejectedTenantsWithAttributes: false,
  errorTenantsWithAttributes: null,
  bestFitFlavors: [],
  isFetchingBestFitFlavor: false,
  isRejectedBestFitFlavor: false,
  errorBestFitFlavor: null,
  instancePropertiesRows: [],
  isSettingSecurityGroupsAndBestFitFlavors: true,
  recalculateBestFitFlavorAndSecurityGroup: true
});

export default (state = initialState, action) => {
  switch (action.type) {
    case `${QUERY_V2V_OSP_TENANT_ATTRIBUTES}_PENDING`:
      return state
        .set('isSettingSecurityGroupsAndBestFitFlavors', true)
        .set('isFetchingTenantsWithAttributes', true)
        .set('isRejectedTenantsWithAttributes', false);
    case `${QUERY_V2V_OSP_TENANT_ATTRIBUTES}_FULFILLED`:
      return state
        .set('tenantsWithAttributes', action.payload.data.results)
        .set('isFetchingTenantsWithAttributes', false)
        .set('isRejectedTenantsWithAttributes', false)
        .set('errorTenantsWithAttributes', null);
    case `${QUERY_V2V_OSP_TENANT_ATTRIBUTES}_REJECTED`:
      return state
        .set('errorTenantsWithAttributes', action.payload)
        .set('isFetchingTenantsWithAttributes', false)
        .set('isRejectedTenantsWithAttributes', true);
    case `${QUERY_V2V_OSP_BEST_FIT_FLAVOR}_PENDING`:
      return state.set('isFetchingBestFitFlavor', true).set('isRejectedBestFitFlavor', false);
    case `${QUERY_V2V_OSP_BEST_FIT_FLAVOR}_FULFILLED`:
      return state
        .set('bestFitFlavors', action.payload.data.results)
        .set('isFetchingBestFitFlavor', false)
        .set('isRejectedBestFitFlavor', false)
        .set('errorBestFitFlavor', null);
    case `${QUERY_V2V_OSP_BEST_FIT_FLAVOR}_REJECTED`:
      return state
        .set('errorBestFitFlavor', action.payload)
        .set('isFetchingBestFitFlavor', false)
        .set('isRejectedBestFitFlavor', true);
    case SET_V2V_BEST_FIT_FLAVORS_AND_DEFAULT_SECURITY_GROUPS: {
      const { vmBestFitFlavors, editingPlan } = action.payload;
      const instancePropertiesRowsUpdatedWithBestFlavor = [];
      vmBestFitFlavors.forEach(vmFlavor => {
        const existingInstancePropertiesRow = state.instancePropertiesRows.find(row => row.id === vmFlavor.vm_id);
        const tenant = state.tenantsWithAttributes.find(
          tenantsWithAttribute => tenantsWithAttribute.id === vmFlavor.tenant_id
        );
        const tenantFlavors = tenant && tenant.flavors;
        const tenantSecurityGroups = tenant && tenant.security_groups;

        let bestFitFlavor;
        let bestFitFlavorId = vmFlavor.flavor_id;
        if (bestFitFlavorId) {
          bestFitFlavor = tenantFlavors && tenantFlavors.find(flavor => flavor.id === bestFitFlavorId);
        } else {
          bestFitFlavor =
            tenantFlavors &&
            tenantFlavors.reduce((prev, current) => {
              if (prev.memory > current.memory) return prev;
              if (prev.memory < current.memory) return current;
              return prev.cpus > current.cpus ? prev : current;
            });
          bestFitFlavorId = bestFitFlavor && bestFitFlavor.id;
        }

        const bestFitFlavorName = bestFitFlavor && bestFitFlavor.name;
        const defaultSecurityGroup = tenant && tenant.default_security_group;
        const defaultSecurityGroupId = defaultSecurityGroup && defaultSecurityGroup.id;
        const defaultSecurityGroupName = defaultSecurityGroup && defaultSecurityGroup.name;

        const bestFlavor = { name: bestFitFlavorName, id: bestFitFlavorId };
        const bestGroup = { name: defaultSecurityGroupName, id: defaultSecurityGroupId };

        let preselectedFlavor;
        let csvInvalidFlavorWarning = false;
        let preselectedGroup;
        let csvInvalidGroupWarning = false;

        if (existingInstancePropertiesRow.csvFields) {
          const { osp_flavor, osp_security_group } = existingInstancePropertiesRow.csvFields;
          if (osp_flavor) {
            const matchingFlavor = tenantFlavors && tenantFlavors.find(flavor => flavor.name === osp_flavor);
            preselectedFlavor = matchingFlavor && { name: matchingFlavor.name, id: matchingFlavor.id };
            csvInvalidFlavorWarning = !matchingFlavor;
          }
          if (osp_security_group) {
            const matchingGroup =
              tenantSecurityGroups && tenantSecurityGroups.find(group => group.name === osp_security_group);
            preselectedGroup = matchingGroup && { name: matchingGroup.name, id: matchingGroup.id };
            csvInvalidGroupWarning = !matchingGroup;
          }
        }

        if (editingPlan) {
          const existingVm = editingPlan.options.config_info.actions.find(
            vm => vm.vm_id === existingInstancePropertiesRow.id
          );
          if (existingVm) {
            const { csvFields } = existingInstancePropertiesRow;
            if (!csvFields || !csvFields.osp_flavor) {
              const existingFlavor =
                tenantFlavors && tenantFlavors.find(flavor => flavor.id === existingVm.osp_flavor_id);
              preselectedFlavor = existingFlavor && { name: existingFlavor.name, id: existingFlavor.id };
            }
            if (!csvFields || !csvFields.osp_security_group) {
              const existingGroup =
                tenantSecurityGroups &&
                tenantSecurityGroups.find(group => group.id === existingVm.osp_security_group_id);
              preselectedGroup = existingGroup && { name: existingGroup.name, id: existingGroup.id };
            }
          }
        }

        const rowUpdatedWithBestFlavor = {
          ...existingInstancePropertiesRow,
          osp_flavor: preselectedFlavor || bestFlavor,
          csvInvalidFlavorWarning,
          osp_security_group: preselectedGroup || bestGroup,
          csvInvalidGroupWarning,
          target_cluster_name: tenant.name
        };
        instancePropertiesRowsUpdatedWithBestFlavor.push(rowUpdatedWithBestFlavor);
      });
      return state
        .set('isSettingSecurityGroupsAndBestFitFlavors', false)
        .set('instancePropertiesRows', instancePropertiesRowsUpdatedWithBestFlavor);
    }
    case SET_V2V_INSTANCE_PROPERTIES_ROWS:
      return state.set('instancePropertiesRows', action.payload);
    case PLAN_WIZARD_EXITED:
      return initialState;
    case PLAN_WIZARD_NEXT:
      return state.set('recalculateBestFitFlavorAndSecurityGroup', action.payload === stepIDs.vmStep);
    default:
      return state;
  }
};
