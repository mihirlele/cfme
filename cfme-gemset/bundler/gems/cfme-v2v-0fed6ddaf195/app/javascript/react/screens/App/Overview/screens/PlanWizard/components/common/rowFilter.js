/**
 * Simple client side row filter for VM table
 *
 * activeFilters: {
 *  label: 'VM Name: {value}',
 *  field: {id: "name", title: "VM Name", placeholder: "Filter by VM Name", filterType: "text"},
 *  value: '{value}'
 * }
 *
 * rows: [{
 *  name: 'cfmetest67',
 *  path: 'vCenter/Datacenter',
 *  cluster: 'Raleigh'
 *  ...
 * }]
 */
export default function rowFilter(activeFilters, rows) {
  if (activeFilters && activeFilters.length && rows && rows.length) {
    const filteredRows = [];
    rows.forEach(row => {
      const match = activeFilters.every(
        filter =>
          row[filter.field.id] && filter.value
            ? `${row[filter.field.id].toLowerCase()}`.indexOf(filter.value.toLowerCase()) > -1
            : null
      );
      if (match) {
        filteredRows.push(row);
      }
    });
    return filteredRows;
  }
  return rows;
}
