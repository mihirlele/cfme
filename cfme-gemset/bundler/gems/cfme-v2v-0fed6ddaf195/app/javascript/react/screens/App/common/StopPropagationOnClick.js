import React from 'react';
import PropTypes from 'prop-types';
import { noop } from 'patternfly-react';

// This component has no role on the page, and will be hidden from screen readers (role="presentation").
// Prevents a click on an inner element (e.g. a DropdownKebab) from triggering a click on an ancestor (e.g. a table row).
// The onKeyUp handler and role are used to satisfy eslint-plugin-jsx-a11y.
const StopPropagationOnClick = ({ children }) => (
  <span onClick={e => e.stopPropagation()} onKeyUp={noop} role="presentation">
    {children}
  </span>
);

StopPropagationOnClick.propTypes = {
  children: PropTypes.node
};

export default StopPropagationOnClick;
