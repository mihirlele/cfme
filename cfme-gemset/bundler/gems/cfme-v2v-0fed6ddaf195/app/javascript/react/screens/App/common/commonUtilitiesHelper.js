const commonUtilitiesHelper = {
  getMostRecentEntityByCreationDate: entities =>
    entities && entities.reduce((prev, current) => (prev.created_on > current.created_on ? prev : current)),

  groupBy: (items, key) =>
    items &&
    items.reduce(
      (result, item) => ({
        ...result,
        [item[key]]: [...(result[item[key]] || []), item]
      }),
      {}
    ),

  flatten: (arrayItems, flattenedArray) =>
    arrayItems && arrayItems.map(task => task.map(arrayElement => flattenedArray.push(arrayElement)))
};

export default commonUtilitiesHelper;
