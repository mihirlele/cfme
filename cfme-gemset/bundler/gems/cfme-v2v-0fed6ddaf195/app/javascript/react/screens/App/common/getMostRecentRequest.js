const getMostRecentRequest = requests =>
  requests.length && requests.reduce((prev, current) => (prev.created_on > current.created_on ? prev : current));

export default getMostRecentRequest;
