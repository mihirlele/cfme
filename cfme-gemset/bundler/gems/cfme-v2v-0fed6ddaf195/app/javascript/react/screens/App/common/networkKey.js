/**
 * returns a unique network based on current UI model
 */
export const networkKey = network => `${network.clusterId}\\${network.uid_ems || network.ems_ref}`;
