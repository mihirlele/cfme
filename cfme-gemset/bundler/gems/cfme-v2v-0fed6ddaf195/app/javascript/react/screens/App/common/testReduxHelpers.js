import thunk from 'redux-thunk';
import promiseMiddleware from 'redux-promise-middleware';
import { createStore, applyMiddleware, combineReducers } from 'redux';
import configureMockStore from 'redux-mock-store';

const middlewares = [thunk, promiseMiddleware()];

export const generateStore = (reducers, initialState) =>
  createStore(combineReducers(reducers), initialState, applyMiddleware(...middlewares));

export const mockStore = initialState => configureMockStore(middlewares)(initialState);
