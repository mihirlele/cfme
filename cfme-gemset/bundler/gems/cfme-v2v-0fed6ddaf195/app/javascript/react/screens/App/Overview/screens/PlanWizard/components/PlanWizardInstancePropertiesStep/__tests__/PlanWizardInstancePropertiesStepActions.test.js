import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import promiseMiddleware from 'redux-promise-middleware';
import * as actions from '../PlanWizardInstancePropertiesStepActions';
import {
  queryTenantsWithAttributesData,
  bestFitFlavorData,
  flavorMappings,
  processedBestFitFlavors
} from '../planWizardInstancePropertiesStep.fixtures';
import { initialState } from '../PlanWizardInstancePropertiesStepReducer';
import { mockRequest, mockReset } from '../../../../../../../../../common/mockRequests';

const middlewares = [thunk, promiseMiddleware()];
const mockStore = configureMockStore(middlewares);
const store = mockStore(initialState);

afterEach(() => {
  store.clearActions();
  mockReset();
});

describe('PlanWizard instance properties step actions', () => {
  it('should query tenants with attributes and return PENDING and FULFILLED actions', () => {
    const { method, argumentUrl, requestUrl, response } = queryTenantsWithAttributesData;
    mockRequest({
      method,
      url: requestUrl,
      response
    });
    return store
      .dispatch(
        actions.queryTenantsWithAttributesAction(
          argumentUrl,
          ['42000000000003'],
          ['flavors', 'security_groups', 'default_security_group']
        )
      )
      .then(() => {
        expect(global.API.post.mock.calls[global.API.post.mock.calls.length - 1][0]).toBe(requestUrl);
        const storeActions = store.getActions();
        expect(storeActions).toHaveLength(2);
        expect(storeActions[0].type).toBe('QUERY_V2V_OSP_TENANT_ATTRIBUTES_PENDING');
        expect(storeActions[1].type).toBe('QUERY_V2V_OSP_TENANT_ATTRIBUTES_FULFILLED');
        expect(storeActions[1].payload).toEqual(response);
      });
  });

  it('should query best fit flavors and return PENDING, SET... and FULFILLED actions', () => {
    const { method, url, response } = bestFitFlavorData;
    mockRequest({ method, url, response });
    return store.dispatch(actions.bestFitFlavorAction(url, flavorMappings, null)).then(() => {
      const storeActions = store.getActions();
      expect(storeActions).toHaveLength(3);
      expect(storeActions[0].type).toBe('QUERY_V2V_OSP_BEST_FIT_FLAVOR_PENDING');
      expect(storeActions[1].type).toBe('SET_V2V_BEST_FIT_FLAVORS_AND_DEFAULT_SECURITY_GROUPS');
      expect(storeActions[1].payload).toEqual(processedBestFitFlavors);
      expect(storeActions[2].type).toBe('QUERY_V2V_OSP_BEST_FIT_FLAVOR_FULFILLED');
      expect(storeActions[2].payload).toEqual(response);
    });
  });

  it('should dispatch the correct action on instancePropertiesRowsAction', () => {
    store.dispatch(actions.instancePropertiesRowsAction([{ mock: 'row1' }, { mock: 'row2' }]));
    expect(store.getActions()).toMatchSnapshot();
  });
});
