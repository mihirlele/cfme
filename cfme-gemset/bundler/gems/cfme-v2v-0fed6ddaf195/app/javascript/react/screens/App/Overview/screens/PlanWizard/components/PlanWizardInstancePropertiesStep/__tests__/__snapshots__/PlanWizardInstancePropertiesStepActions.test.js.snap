// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanWizard instance properties step actions should dispatch the correct action on instancePropertiesRowsAction 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "mock": "row1",
      },
      Object {
        "mock": "row2",
      },
    ],
    "type": "SET_V2V_INSTANCE_PROPERTIES_ROWS",
  },
]
`;
